Handle LocalDate values in Excel export

robdaemon opened the pull request:
Adds an additional handler for LocalDate objects and a graceful
degradation of any other errors converting date/time objects to Java
Instants.

Resolves #13271
Resolves #10803
Resolves #13298

Commit:
Handle LocalDate values in Excel export

Adds an additional handler for LocalDate objects and a graceful
degradation of any other errors converting date/time objects to Java
Instants.

Resolves #13271
Resolves #10803
Resolves #13298

Commit:
adding a test case for the exceptional case

We don't handle JapaneseDate objects internally, but they make a good
example for what happens in an exceptional case in generating Excel
files.

sbelak reviewed (approved):
Looks good!

flamber reviewed (changes requested):
Current state fixes #13298, but not #13271 and there are some issues with #10803:

1. When export SQL/QB of a saved question, then it still contains the full timezone string for both date and timestamp columns. And sometimes the date-column has a timezone, where the DST doesn't seem to be correct every time (ex People ID=2).
2. When exporting SQL/QB of dirty (unsaved) question, then it correctly converts timestamp-column, but displays it as date only (XLSX NumFmtId=14) instead of date+time. But a date-column is now getting the time of the export added to it's cell, and is also only shown as NumFmtId=14.

Several of these problems also occurs on CSV - like issue #12964

I have just tested with Sample Dataset > People.

Commit:
More updates for #10803

Making sure formatting to strings isn't enabled on CSV and XLSX exports
on cards and datasets.

Update formatting of the Excel document to handle the difference between
dates and date-times.

Commit:
card exports should not convert integers to strings

Commit:
not working, but the general approach to reusing cell styles

Commit:
Use a dynamic in the right place to hold CellStyles

Commit:
appease the linter bots

flamber commented:
Tested `599cc79`.

This now fixes #6012, #10803, #12964, #13298 and #13271 (although a tiny bit slower compared to 0.36.4, but that's expected)
Awesomeness :tada: :champagne: 

But it still adds the time of the export to any date columns (either from table or by binning Day on any timestamp-like column).
If we need to hack something, then adding time `00:00:00` would be more correct.
- Simple question > Sample Dataset > People > export as XLSX.
The first row of "Birth Date" will be shown as `12/12/1986`, but the cell value is `12/12/1986 11:45:36` (where the time will depend on when it was exported)

And a minor thing is that the XLSX column width isn't auto-sized, so it now shows datetime columns as `###` until resized.
![image](https://user-images.githubusercontent.com/1447303/96258984-2fcafd80-0fbd-11eb-89c0-ea270e77d020.png)
But we have #9251 for that.

sbelak commented on src/metabase/query_processor/streaming/xlsx.clj:
```diff
@@ -17,38 +19,68 @@
+                     format-string
+                     (doto date-style
+                       (.setDataFormat (.. format-helper createDataFormat (getFormat format-string))))))]
+      (reset! *cell-styles* new-val)))
```
why `reset!` instead of `swap!` + `assoc`?

robdaemon replied:
:shrug: Is one style preferred over the other?

camsaul replied:
`swap!` is preferable because it performs a transformation atomically while `reset!` replaces the whole thing atomically. It doesn't really matter here because it's non running on multiple threads but `swap!` will do all the updates without stomping on changes from other threads while `reset!` will just completely replace the contents of the atom with no regards to what any other threads might be doing

camsaul replied:
swap! = apply a transformation
reset! = completely reset the contents of the atom

sbelak commented on src/metabase/query_processor/streaming/xlsx.clj:
```diff
@@ -17,38 +19,68 @@
+  "Standard date/time format for any of the :type/Date variants with a Time"
+  "m/d/yy HH:MM:ss")
+
+(defmethod spreadsheet/set-cell! LocalDate [^Cell cell val]
```
These all look super similar. Can we somehow factor the repeating bits out?

robdaemon replied:
They're similar except for the date format, they can likely be refactored. I'll give that a shot.

sbelak commented on src/metabase/query_processor/streaming/xlsx.clj:
```diff
@@ -66,14 +98,16 @@
 
       (write-row! [_ row _]
-        (spreadsheet/add-row! sheet (map excel-format-value row)))
+        (binding [*cell-styles* cell-styles]
```
I know you had a huge debate around this, but if performance is a concern you could also have a global atom that is a map of workbook->cell-styles that way it's still thread-isolated but you don't have to go through the binding mechanism for each row. What I do like about the dynamic solution is that it also implicitly cleans up after itself, while with a global atom we'd have to do that manually (perhaps with a macro for clarity).

sbelak reviewed (approved):
Looks like it fixes a shit ton of bugs, great job!

Commit:
refactor set-cell! duplication

set the time on a LocalDate to midnight UTC, which truncates the decimal
portion of a date-only object to 0 in Excel

robdaemon commented:
> But it still adds the time of the export to any date columns (either from table or by binning Day on any timestamp-like column).
> If we need to hack something, then adding time `00:00:00` would be more correct.

@flamber I addressed this - it now sets any `LocalDate` type to `00:00` on the time, so you'll see a `x.0` value in the XML.

Commit:
replace reset! with swap!

flamber reviewed (approved):
I have tested this several times over, and I think we're good now. This is a big improvement over the old version (pre-0.36.4) and the current version (0.36.7) :gift_heart:

_I might be slightly blinded by testing this so much for several days - I'm XLSX :vomiting_face:_

Commit:
appease the bikeshedder
